spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarConsumerFactory.java

```diff
@@ -16,26 +16,65 @@
 
 package org.springframework.pulsar.core;
 
+import java.util.Collection;
+import java.util.List;
 import java.util.Map;
 
-import org.apache.pulsar.client.api.BatchReceivePolicy;
 import org.apache.pulsar.client.api.Consumer;
 import org.apache.pulsar.client.api.PulsarClientException;
 import org.apache.pulsar.client.api.Schema;
+import org.apache.pulsar.client.impl.conf.ConsumerConfigurationData;
+
+import org.springframework.lang.Nullable;
 
 /**
  * Pulsar consumer factory interface.
  *
  * @param <T> payload type for the consumer.
  * @author Soby Chacko
+ * @author Christophe Bornet
  */
 public interface PulsarConsumerFactory<T> {
 
-	Consumer<T> createConsumer(Schema<T> schema, Map<String, Object> propertiesToOverride) throws PulsarClientException;
+	/**
+	 * Create a consumer.
+	 * @param schema the schema of the messages to be sent
+	 * @return the consumer
+	 * @throws PulsarClientException if any error occurs
+	 */
+	Consumer<T> createConsumer(Schema<T> schema) throws PulsarClientException;
+
+	/**
+	 * Create a consumer.
+	 * @param schema the schema of the messages to be sent
+	 * @param customizers the optional list of customizers to apply to the consumer
+	 * builder
+	 * @return the consumer
+	 * @throws PulsarClientException if any error occurs
+	 */
+	Consumer<T> createConsumer(Schema<T> schema, List<ConsumerBuilderCustomizer<T>> customizers)
+			throws PulsarClientException;
 
-	Consumer<T> createConsumer(Schema<T> schema, BatchReceivePolicy batchReceivePolicy,
-			Map<String, Object> propertiesToOverride) throws PulsarClientException;
+	/**
+	 * Create a consumer.
+	 * @param schema the schema of the messages to be sent
+	 * @param topics the topics the consumer will subscribe to or {@code null} to use the
+	 * default topics
+	 * @param properties the properties to attach to the consumer or {@code null} to use
```

Comment:
I used the word `attach` from the definition in `ConsumerConfigurationData`.
I'll find a better formulation.
```java
    @ApiModelProperty(
            name = "properties",
            value = "A name or value property of this consumer.\n"
                    + "\n"
                    + "`properties` is application defined metadata attached to a consumer.\n"
                    + "\n"
                    + "When getting a topic stats, associate this metadata with the consumer stats for easier "
                    + "identification."
    )
    private SortedMap<String, String> properties = new TreeMap<>();
```

Reply:
👍🏻 I think "replace" or "override" would do the trick (depending on what does of course).

Reply:
**[NIT]** This is still "attach". I think going back to the original name of `propertiesToOverride` will clarify this. 

Also, I just realized that we were doing a "merge/override" before rather than a wholesale "replace". I believe we still want to allow them to override specific props rather than replace the whole map. 

In that case, the `CB.properties` does the right things (via `Map.putAll`) and we should be able to pull the properties param out altogether, no?

Reply:
No that was already a replace. Here, it's the consumer properties as in `ConfigurationConsumerData.properties`, not the properties (previously called `propertiesToOverride `) from which we do `loadConf`.